GlCanvas cleanup

reichlfl opened the pull request:
This is a first refactor of GlCanvas / CaptureWindow. Sorry for the large PR, I kept running into small inconsistencies during the refactor that made it hard to find a reasonable cutoff point, and I wanted to avoid submitting it half-done because I had to revisit and change previous commits quite a bit as my assumptions continued to be incorrect...

### Goals of this PR

**Remove the code duplication and inconsistent behavior of `GlCanvas` and `CaptureWindow`** 

A lot of functionality (e.g. picking) was partially implemented in `GlCanvas`, but only worked for `CaptureWindow`. Other functionality was implemented in `GlCanvas` and then reimplemented (sometimes with minor changes) in `CaptureWindow`. Now, `GlCanvas` fully takes care of picking and tracking mouse interactions, whereas `CaptureWindow` add all functionality related to `TimeGraph` and Scrollbars.

**Break down GlCanvas to improve Unit Test coverage**

As a first step, all functionality related to viewport sizes, coordinate conversion, and virtual "camera placement" (e.g. position of the screen within the 2D world) moved to a new class `Viewport`. This new class is properly unit tested.

### Next steps
* Pass `Batcher*` and `Viewport*` instead of `GlCanvas*` where possible. This will decouple most components from `GlCanvas`, and will decouple rendering from coordinate translation.
* Change horizontal scrolling and scaling to use the `Viewport` functionality. Currently, this is handled by `TimeGraph`, which is quite inconsistent (`Viewport::GetWorldTopLeft()[0]` is always 0!)
* Move the layering definitions into a separate class
* Move the actual "Renderpass" functionality into a new class that is linked to a viewport - so we get rid of the implicit assumptions that "some elements are drawn in screen space, others in world space", and remove the implicit 2nd renderpass that happens on picking, as this causes issues with state changes in between those two renderpasses.

karupayun commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(50.f, -50.f));
+}
+
+TEST(Viewport, ScrollingWithZoom) {
```
Could we also add a test where we zoom in the middle of the test and check if the point_new_position is where it should be?

reichlfl replied:
It should be covered by the coordinate conversion test. I've extended that one to include a zoomed and non-zoomed version of all tests.

karupayun commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+  Viewport viewport(10, 100);
+  // Extents are large enough to scroll
+  viewport.SetWorldExtents(500, 500);
+  viewport.SetWorldWidth(20.f);
```
A small comment here to explain the zoom-level?
Something like this:
// horizontal zoom_level = 2, vertical zoom-level = 0.5?

reichlfl replied:
Done

karupayun commented on src/OrbitGl/GlCanvas.cpp:
```diff
@@ -176,11 +167,16 @@ void GlCanvas::MouseMoved(int x, int y, bool left, bool /*right*/, bool /*middle
-                        static_cast<float>(mouse_x) / static_cast<float>(GetWidth()) * world_width_;
-    world_top_left_y_ = world_click_y_ + static_cast<float>(mouse_y) /
-                                             static_cast<float>(GetHeight()) * world_height_;
+  if (left && !picking_manager_.IsDragging()) {
```
Why the dragging here? Were there a bug?

reichlfl replied:
`picking_manager_.IsDragging() == true` means that the user clicked on an element in the capture window and is dragging across it. In this case, there should be no panning happening in GlCanvas, the current element will handle any drag-related actions. This is the same logic as before, it was just moved from `CaptureWindow` to `GlCanvas`.

karupayun reviewed (commented):
I'm only commenting the first commit here. In the second commit, everything seems to makes sense, but I would like to review the GlCanvas part.

pierricgimmig reviewed (commented):
Hi @reichlfl , thanks a lot for this, looks really good. I love to see all this red in the diff view :-). Please see comments. Also, I tested the branch and I have found a few bugs:

- Green selection rectangle is not sticky anymore (is this intended?). It is an important feature for the upcoming selection stats.
- Right click used to give Orbit focus, now if Orbit is not in focus and we try to right-click and drag, the green rectangle does not appear
- Vertical scrolling in a single direction makes the vertical slider jump up and down (see video)


https://user-images.githubusercontent.com/3687222/113441777-baed5380-93a3-11eb-972d-ecd50b5aa217.mov

pierricgimmig commented on src/OrbitGl/GlCanvas.cpp:
```diff
@@ -241,29 +195,35 @@ void GlCanvas::LeftUp() {
   Orbit_ImGui_MouseButtonCallback(imgui_context_, 1, false);
-  is_selecting_ = true;
+  is_selecting_ = false;
+  select_start_ = select_stop_;
```
This line makes the green selection rectangle disappear on "right-up" which is a change in behavior. I think we should keep the previous behavior.

reichlfl replied:
Thanks for catching this, fixed now!

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+
+namespace orbit_gl {
+
+/**
```
Nit: If I am not wrong we avoid block comments and use `//` at all places. However, even if not it should be
```
/*
 *
 */
```

reichlfl replied:
I've changed it to //

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  explicit Viewport(int width, int height);
+
+  void Resize(int width, int height);
+  [[nodiscard]] int GetWidth() const;
```
When reading this and the class level comment, I am wondering if we should make it more clear, that those values refer to the screen, (e.g. `GetScreenWidth`, as for the world below).

reichlfl replied:
Yeah, true, I hope it's now clearer from the comments and the renaming of the methods.

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  void SetWorldExtents(float width, float height);
+  [[nodiscard]] Vec2 GetWorldExtents();
+
+  void SetWorldTopLeftY(float y);
```
Will they be called individually? Otherwise, consider passing a vec. to it.

Also I think there should be a comment.

reichlfl replied:
I thought about that too for the sake of API consistency, but these are called individually quite often (e.g. when manipulating through the scrollbars).

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  [[nodiscard]] int GetWidth() const;
+  [[nodiscard]] int GetHeight() const;
+
+  // Visible size of the world
```
Nit: Add `.` and below.
https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

reichlfl replied:
Done.

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  [[nodiscard]] const Vec2& GetWorldTopLeft() const;
+
+  [[nodiscard]] Vec2 ScreenToWorldPos(const Vec2i& screen_pos) const;
+  [[nodiscard]] float ScreenToWorldHeight(int height) const;
```
`Height` or `Y`? Same for the width.

reichlfl replied:
Added a comment here. There's a difference between Pos and Width / Height, that's why we need both methods.

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+
+  [[nodiscard]] Vec2i QtToGlScreenPos(const Vec2i& qt_pos) const;
+
+  [[nodiscard]] bool IsDirty() const { return is_dirty_; }
```
Could you add a comment here? To me it is not clear what `Dirty` means in this context.

reichlfl replied:
Done

florian-kuebler commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
+
+  // Test initial values
```
Nit: Consider adding punctation to the comments here.

reichlfl replied:
Done

florian-kuebler commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+TEST(Viewport, ResizingAndDirty) {
+  Viewport viewport(100, 200);
+  EXPECT_TRUE(viewport.IsDirty());
+  viewport.ClearDirtyFlag();
```
Also add an `EXPECT_FALSE(viewport.IsDirty()` afterwards?

reichlfl replied:
Makes sense, thanks. Added.

florian-kuebler commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+
+where world(0, 0) is initially anchored at screen(0, 0), i.e. top left (meaning basically all world
+y coordinates are negative). Viewport clamps all X values to be >= 0, and all Y values to be <= 0.
+
```
I would also expect at least some very brief comments about the usage (e.g. that the ClearDirtyFlag needs to be called).

reichlfl replied:
Done

florian-kuebler commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+  EXPECT_EQ(viewport.GetWorldWidth(), 500.f);
+  EXPECT_EQ(viewport.GetWorldHeight(), 600.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(0.f, 0.f));
+  EXPECT_EQ(viewport.GetWorldExtents(), Vec2(100, 200));
```
I agree with this being not super easy to grasp at the first glance. Maybe consider adding a more detailed comment on the method (IDE's should show the comments also on the caller).

reichlfl replied:
Good point. I was thinking the same and was considering to rename `WorldWidth` to `WorldVisibleWidth` or something, and decided against it to make it easier to review (as the naming doesn't change so much). But I think I'll actually add the renaming to this PR.

reichlfl replied:
I renamed all methods:

VisibleWorld[Width/Height]: Size of the world visible on the screen
WorldExtents: Max values of the current world
WorldMin: Min values of the current world (*technically*, min and max are swapped for y because all values we use are negative, but conceptually, I think I'll stick with this notion of min / max)
ScreenSize: Size of the screen in pixels

I added some comments above each method block, hope that makes the usage clearer.

florian-kuebler commented on src/OrbitGl/Viewport.cpp:
```diff
@@ -0,0 +1,130 @@
+
+namespace orbit_gl {
+
+Viewport::Viewport(int width, int height) : width_(width), height_(height) {
```
Nit: also `world_width_(width), world_height_(height)`?

reichlfl replied:
And extents_, actually :-) Changed.

florian-kuebler commented on src/OrbitGl/Viewport.cpp:
```diff
@@ -0,0 +1,130 @@
+  FlagAsDirty();
+}
+
+int Viewport::GetWidth() const { return width_; }
```
Nit: Those one-liners can be moved to the header.

reichlfl replied:
I prefer to have them in the cpp - I assume this class will be included quite a lot at some point, and we already suffer from compilation times whenever we change anything in GlCanvas.h, so I'd like to keep the potential for header changes minimal here.

reichlfl commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+    v
+
+where world(0, 0) is initially anchored at screen(0, 0), i.e. top left (meaning basically all world
+y coordinates are negative). Viewport clamps all X values to be >= 0, and all Y values to be <= 0.
```
IMO we shouldn't, because it's part of the Viewport's functionality to do this, and happened in other places in GlCanvas / CaptureWindow before. The unit tests explicitly check for this as intended behavior.
I could see value in making the min-value (currently always 0) configurable, then this would be the equivalent to worldExtents, WDYT?

karupayun replied:
I like it! But if it isn't trivial, don't rush in making it for this PR.

reichlfl replied:
Done, see comment below.

florian-kuebler commented on src/OrbitGl/Viewport.cpp:
```diff
@@ -0,0 +1,130 @@
+  if (width == world_width_) return;
+
+  world_width_ = width;
+  // Recalculate required scrolling
```
Nit: Add `.` and below.

reichlfl replied:
Done

florian-kuebler commented on src/OrbitGl/ThreadTrack.cpp:
```diff
@@ -276,17 +276,17 @@ void ThreadTrack::Draw(GlCanvas* canvas, PickingMode picking_mode, float z_offse
 
   if (!thread_state_bar_->IsEmpty()) {
-    thread_state_bar_->SetSize(canvas->GetWorldWidth(), thread_state_track_height);
+    thread_state_bar_->SetSize(canvas->GetViewport().GetWorldWidth(), thread_state_track_height);
```
Nit: Extract the world_width into a local var.?

reichlfl replied:
Done

reichlfl commented on src/OrbitGl/CaptureWindow.cpp:
```diff
@@ -642,13 +470,12 @@ void CaptureWindow::UpdateHorizontalScroll(float ratio) {
   if (time_graph_ == nullptr) return;
-  float min = world_max_y_;
-  float max = world_height_ - time_graph_->GetTrackManager()->GetTracksTotalHeight();
+  float min = 0.0f;
```
`world_max_y_` was 0 before since all coordinates in the world have negative y. This is now managed by the `Viewport`, but maybe it makes sense to retrieve the max-value from there instead of hardcoding it.

karupayun replied:
Yes, or at least a comment to explain why it is 0

reichlfl replied:
Done

reichlfl commented:
> * Green selection rectangle is not sticky anymore (is this intended?). It is an important feature for the upcoming selection stats.
> * Right click used to give Orbit focus, now if Orbit is not in focus and we try to right-click and drag, the green rectangle does not appear
> * Vertical scrolling in a single direction makes the vertical slider jump up and down (see video)

The intent was to keep the existing behavior, so all changes mentioned above can be considered bugs. I'll take a look and fix them.

reichlfl commented on src/OrbitGl/Viewport.cpp:
```diff
@@ -0,0 +1,130 @@
+
+  width_ = width;
+  height_ = height;
+  FlagAsDirty();
```
Yes, any action that changes the visible or total area of the world. Basically anything that would require `UpdatePrimitives` on the time graph.

reichlfl replied:
I've added a comment and a brief explanation in the file header.

florian-kuebler commented on src/OrbitGl/CaptureWindow.cpp:
```diff
@@ -247,276 +188,161 @@ void CaptureWindow::SelectTextBox(const TextBox* text_box) {
-  select_stop_ = select_start_;
-  time_start_ = time_graph_->GetTickFromWorld(world_click_x_);
-  time_stop_ = time_start_;
+  // TODO: This should not be needed, but the extents of track heights may change after drawing
```
Consider creating a bug for this.

reichlfl replied:
This was moved into `CaptureWindow::Draw` now and the comment was removed.

florian-kuebler commented on src/OrbitGl/CaptureWindow.cpp:
```diff
@@ -675,25 +502,17 @@ void CaptureWindow::UpdateHorizontalSliderFromWorld() {
-  float max = world_height_ - time_graph_->GetTrackManager()->GetTracksTotalHeight();
-  float ratio = (world_top_left_y_ - min) / (max - min);
-  float vertical_ratio = world_height_ / time_graph_->GetTrackManager()->GetTracksTotalHeight();
+  float min = 0.0f;
```
Same as Pierric mentioned above.

reichlfl replied:
Done

florian-kuebler commented on src/OrbitGl/GlCanvas.cpp:
```diff
@@ -1,16 +1,16 @@
+#include "GlCanvas.h"
```
The licence declaration should be the first thing in a file:
https://google.github.io/styleguide/cppguide.html#File_Comments

florian-kuebler replied:
Also, it seems that you are including `GlCanvas` twice.

reichlfl replied:
Hm, seems that clang-format messed that up, thanks for catching.

reichlfl commented on src/OrbitGl/Viewport.cpp:
```diff
@@ -0,0 +1,130 @@
+Vec2 Viewport::GetWorldExtents() { return world_extents_; }
+
+void Viewport::SetWorldTopLeftY(float y) {
+  float clamped = std::min(std::max(y, world_height_ - world_extents_[1]), 0.f);
```
I've decided against this for 2 reasons:
1) I want the user of the class to be able to zoom out to more than 100% - for which you need to set world height > world extents
2) I don't want changes in the visible height to affect the extents because this could lead to unwanted side-effects.

I couldn't use `std::clamp` unfortunately since it doesn't deal with the case `world_extents_[1] < world_height_` correctly.

karupayun replied:
I see your point now (and actually I think I modified it in the past when I made the vertical zoom and thought the same). My plan seems to be clearly more dirty and bug-prone. Thanks for the explanation :)

karupayun commented on src/OrbitGl/GlCanvas.h:
```diff
@@ -73,18 +72,14 @@ class GlCanvas {
   [[nodiscard]] TextRenderer& GetTextRenderer() { return text_renderer_; }
 
-  // TODO: Only needed for Graph labels
+  // Only needed for Graph labels - should we find a better way?
```
We should create a bug for that

reichlfl replied:
I've actually removed this, the method now exposes the last mouse pos (x and y), that makes it a bit more generically useful.

reichlfl commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -0,0 +1,176 @@
+  viewport.SetWorldTopLeftX(350.f);
+  viewport.SetWorldTopLeftY(-600.f);
+  EXPECT_EQ(viewport.GetWorldTopLeft(), Vec2(200.f, -400.f));
+  EXPECT_EQ(viewport.IsDirty(), false);
```
I would leave it in - it tests the same behavior, but the preconditions are different, and this test makes sure that, when zooming is present, the clamping still works as expected.

reichlfl commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  void ClearDirtyFlag() { is_dirty_ = false; }
+
+ private:
+  int width_;
```
Good idea, done, thanks!

reichlfl commented on src/OrbitGl/AccessibleThreadBar.cpp:
```diff
@@ -46,9 +46,10 @@ orbit_accessibility::AccessibilityRect orbit_gl::AccessibleThreadBar::Accessible
+      canvas->GetViewport().WorldToScreenWidth(local_pos[0]),
+      canvas->GetViewport().WorldToScreenHeight(local_pos[1]),
+      canvas->GetViewport().WorldToScreenWidth(thread_bar_->GetSize()[0]),
+      canvas->GetViewport().WorldToScreenHeight(thread_bar_->GetSize()[1]));
```
I've done that where appropriate now. At some point, this can probably disappear anyways because we should try to pass the Viewport instead of GlCanvas where possible.

reichlfl commented on src/OrbitGl/AccessibleTimeGraph.cpp:
```diff
@@ -18,7 +18,8 @@ using orbit_accessibility::AccessibleInterface;
   GlCanvas* canvas = time_graph_->GetCanvas();
-  return AccessibilityRect(0, 0, canvas->GetWidth(), canvas->GetHeight());
+  return AccessibilityRect(0, 0, canvas->GetViewport().GetWidth(),
+                           canvas->GetViewport().GetHeight());
```
Done

reichlfl commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -106,10 +106,10 @@ AccessibilityRect AccessibleTrackContent::AccessibleLocalRect() const {
+  return AccessibilityRect(canvas->GetViewport().WorldToScreenWidth(0),
+                           canvas->GetViewport().WorldToScreenHeight(layout_->GetTrackTabHeight()),
+                           canvas->GetViewport().WorldToScreenWidth(track_->GetSize()[0]),
+                           canvas->GetViewport().WorldToScreenHeight(track_->GetSize()[1]));
```
Done

reichlfl commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -136,10 +136,10 @@ AccessibilityRect AccessibleTrackTab::AccessibleLocalRect() const {
+  return AccessibilityRect(canvas->GetViewport().WorldToScreenWidth(layout_->GetTrackTabOffset()),
+                           canvas->GetViewport().WorldToScreenHeight(0),
+                           canvas->GetViewport().WorldToScreenWidth(layout_->GetTrackTabWidth()),
+                           canvas->GetViewport().WorldToScreenHeight(layout_->GetTrackTabHeight()));
```
Done

reichlfl commented on src/OrbitGl/AccessibleTrack.cpp:
```diff
@@ -175,17 +175,18 @@ AccessibilityRect AccessibleTrack::AccessibleLocalRect() const {
-  int screen_height = canvas->WorldToScreenHeight(size[1]);
+  Vec2i screen_pos = canvas->GetViewport().WorldToScreenPos(pos);
+  int screen_width = canvas->GetViewport().WorldToScreenWidth(size[0]);
+  int screen_height = canvas->GetViewport().WorldToScreenHeight(size[1]);
```
Done

reichlfl commented on src/OrbitGl/AccessibleTriangleToggle.cpp:
```diff
@@ -17,16 +17,16 @@ orbit_accessibility::AccessibilityRect AccessibleTriangleToggle::AccessibleLocal
-  int screen_height = canvas->WorldToScreenHeight(size[1]);
+  Vec2i screen_pos = canvas->GetViewport().WorldToScreenPos(pos);
+  int screen_width = canvas->GetViewport().WorldToScreenWidth(size[0]);
+  int screen_height = canvas->GetViewport().WorldToScreenHeight(size[1]);
```
Done

reichlfl commented on src/OrbitGl/TextRenderer.cpp:
```diff
@@ -297,13 +297,16 @@ void TextRenderer::AddText(const char* text, float x, float y, float z, const Co
+    (*out_text_size)[0] =
+        canvas_->GetViewport().ScreenToWorldWidth(static_cast<int>(out_screen_size.x));
+    (*out_text_size)[1] =
+        canvas_->GetViewport().ScreenToWorldHeight(static_cast<int>(out_screen_size.y));
```
Done

reichlfl commented on src/OrbitGl/TextRenderer.cpp:
```diff
@@ -433,17 +436,17 @@ std::vector<float> TextRenderer::GetLayers() const {
-  o_x = ((x - world_top_left_x) / world_width) * canvas_->GetWidth();
-  o_y = ((y - world_min_left_y) / world_height) * canvas_->GetHeight();
+  o_x = ((x - world_top_left_x) / world_width) * canvas_->GetViewport().GetWidth();
+  o_y = ((y - world_min_left_y) / world_height) * canvas_->GetViewport().GetHeight();
```
Done

reichlfl commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+  (0, 0) ---->
+    |
+    |
+    v
```
Done

reichlfl commented on src/OrbitGl/Viewport.h:
```diff
@@ -0,0 +1,85 @@
+See TODO(b/177350599): Unify QtScreen and GlScreen
+**/
+class Viewport {
+ public:
+  explicit Viewport(int width, int height);
```
Yes, good point, done!

reichlfl commented on src/OrbitQt/orbitglwidget.cpp:
```diff
@@ -190,7 +190,10 @@ void OrbitGLWidget::messageLogged(const QOpenGLDebugMessage& msg) {
   if (gl_canvas_) {
     gl_canvas_->Resize(w, h);
-    gl_canvas_->SetMainWindowSize(this->geometry().width(), this->geometry().height());
+    // TODO: I've added this to make sure removing the "MainWindowSize" won't break anything
```
Removed

reichlfl commented on src/OrbitGl/GlCanvas.cpp:
```diff
@@ -275,26 +262,14 @@ void GlCanvas::KeyPressed(unsigned int key_code, bool ctrl, bool shift, bool alt
 void GlCanvas::KeyReleased(unsigned int key_code, bool ctrl, bool shift, bool alt) {
   UpdateSpecialKeys(ctrl, shift, alt);
   Orbit_ImGui_KeyCallback(imgui_context_, static_cast<int>(key_code), false, ctrl, shift, alt);
+
```
Done

reichlfl commented on src/OrbitGl/GlCanvas.h:
```diff
@@ -95,14 +88,13 @@ class GlCanvas {
+
   void UpdateWorldTopLeftY() { UpdateWorldTopLeftY(GetWorldTopLeftY()); }
   virtual void UpdateWorldTopLeftY(float val) { world_top_left_y_ = val; }
+  virtual void UpdateWorldTopLeftX(float val) { world_top_left_x_ = val; }
```
I think this is a perfectly valid complaint :D This commit is outdated, but I adjusted it accordingly in Viewport.h.

reichlfl commented:
Thanks for the thorough reviews, @pierricgimmig @karupayun @florian-kuebler. I've addressed the comments in follow-up commits, PTAL.

Pierric: I did not yet find the reason for the weird scrollbar jumping, but I'll investigate further before merging this. I'll update as soon as I find something.

pierricgimmig reviewed (approved):
Cool, thanks for the changes @reichlfl , LGTM, cheers!

reichlfl commented:
> * Right click used to give Orbit focus, now if Orbit is not in focus and we try to right-click and drag, the green rectangle does not appear

@pierricgimmig On windows, this works as expected... sure this is related to the changes here? I did not touch anything above OrbitGl, and can't see how those chances should influence window focus as all of this handling is done in OrbitQt.

> * Vertical scrolling in a single direction makes the vertical slider jump up and down (see video)

@pierricgimmig I've found the reason and fixed it. This was actually a one-line change, I moved the update of Viewport Extents to be done right after track rendering.

karupayun commented on src/OrbitGl/ViewportTest.cpp:
```diff
@@ -151,20 +164,26 @@ void VerifyConversion(Viewport& viewport, const Vec2i& screen_pos, const Vec2& w
+  Vec2 world_size = Vec2(8.f, 20.f);
+  VerifyConversion(viewport, screen_pos, world_pos, world_size);
+
+  // Change zoom: Zoom out 100% horizontally, zoom in 50% vertically.
```
200% horizontally

reichlfl replied:
*By* 100% was what I wanted to say here. But this is not clear, I'll change the comment, thanks!

karupayun replied:
I've just seen "zoom-out". Still prefers 200% and 50%.

reichlfl replied:
Changed

karupayun reviewed (approved):
Only a small nit, otherwise LGTM!!! :)

Thanks for this
